### PR TITLE
Catalog conditions blade with support multiple selection products.

### PR DESCRIPTION
**Pull request by blazemachine**

### CODE REVIEW

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-conditions.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+angular.module('virtoCommerce.marketingModule')
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-conditions.js]: Can we name this blade more precise, like "Catalog-entry-list"?

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js [hunk 2]
```diff
@@ -22,52 +22,165 @@ angular.module('virtoCommerce.marketingModule')
...
+                        catalogBlade.error = undefined;
+                    }
+                }
+            };
+
+            bladeNavigationService.showBlade(catalogBlade, $scope.blade);
+            return;
+        }
+
+        var openCatalogBlade = function (catalogConditionsBlade, parentBlade) {
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Could we make not inlined function for opening catalog entry select blade?

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js [hunk 3]
```diff
@@ -22,52 +22,165 @@ angular.module('virtoCommerce.marketingModule')
...
+                id: "CatalogEntrySelect",
+                title: "marketing.blades.catalog-items-select.title-product",
+                controller: 'virtoCommerce.catalogModule.catalogItemSelectController',
+                template: 'Modules/$(VirtoCommerce.Catalog)/Scripts/blades/common/catalog-items-select.tpl.html',
+                breadcrumbs: [],
+                toolbarCommands: [
+                {
+                    name: "platform.commands.pick-selected", icon: 'fa fa-plus',
+                    executeMethod: function (blade) {
+                        blade.parentBlade.$scope.gridApi.grid.options.data = _.map(blade.parentBlade.selectedListEntries, function (entry) {
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Probably better to make all work with grid in blade were grid is located and created. As i understand it is 'catalog-entry-list' blade responsibility to add, show or delete items in the list. You could place all these methods in this blade.

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js [hunk 4]
```diff
@@ -22,52 +22,165 @@ angular.module('virtoCommerce.marketingModule')
...
-                    return selectedListEntries.length == 1;
+                    return true;
                 }
-            }]
-        };
+            },
+            {
+                name: "platform.commands.delete", icon: 'fa fa-trash-o',
+                executeMethod: function (blade) {
+                    var grid = blade.$scope.gridApi.grid;
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Probably better to make all work with grid in blade were grid is located and created.

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js [hunk 5]
```diff
@@ -22,52 +22,165 @@ angular.module('virtoCommerce.marketingModule')
...
+                    });
+                    blade.productIds = parentElement.productIds.slice();
+                },
+                canExecuteMethod: function (blade) {
+                    if (blade.childrenBlades.length > 0) {
+                        return false;
                     }
-                    else {
-                        selectedListEntries = _.reject(selectedListEntries, function (x) { return x.id == listItem.id; });
+                    if (blade.$scope.gridApi) {
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Probably better to make all work with grid in blade were grid is located and created.

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js [hunk 6]
```diff
@@ -22,52 +22,165 @@ angular.module('virtoCommerce.marketingModule')
...
-                    newBlade.error = undefined;
+                    return false;
                 }
-            }
+            }]
+        };
+
+        newBlade.refresh = function () {
+            newBlade.isLoading = true;
+            items.query({ ids: newBlade.productIds, respGroup: 'ItemInfo' }, function (data) {
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Probably better to make all work with grid in blade were grid is located and created.

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js [hunk 7]
```diff
@@ -118,5 +231,5 @@ angular.module('virtoCommerce.marketingModule')
         };
 
         bladeNavigationService.showBlade(newBlade, $scope.blade);
-    }
+    };
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Is it needed?

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js [hunk 8]
```diff
@@ -22,52 +22,165 @@ angular.module('virtoCommerce.marketingModule')
...
+                newBlade.selectedListEntries = newBlade.$scope.gridApi.grid.options.data.slice();
+                newBlade.isLoading = false;
+            });
         };
 
+        newBlade.productIds = parentElement.productIds != undefined ? parentElement.productIds : [];
+        newBlade.selectedListEntries = selectedListEntries;
         bladeNavigationService.showBlade(newBlade, $scope.blade);
-    }
+    };
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Is is needed?

Comment (artem-dudarev) [reply VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Yes, this is assignment, not just a function definition.

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    .controller('virtoCommerce.marketingModule.catalogEntriesController', ['$scope', 'virtoCommerce.catalogModule.items', 'platformWebApp.ui-grid.extension', 'platformWebApp.bladeUtils', 'platformWebApp.uiGridHelper', 'platformWebApp.bladeNavigationService',
+        function ($scope, items, gridOptionExtension, bladeUtils, uiGridHelper, bladeNavigationService) {
+            var blade = $scope.blade;
+
+            $scope.isEntityChanged = function () {
+                return angular.equals(this.blade.productIds, this.blade.currentProductIds);
+            };
+
+            $scope.saveChanges = function () {
+                this.blade.promotion.productIds = this.blade.productIds.slice();
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: Do we need to make slice here?

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            blade.refresh = function () {
+                initialize();
+            };
+
+            $scope.cancel = function () {
+                initialize();
+            };
+
+            function initialize() {
+                blade.productIds = blade.promotion.productIds != undefined ? blade.promotion.productIds.slice() : [];
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: "blade.promotion.productIds != undefined" could be simplified with "blade.promotion.productIds"

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                blade.currentProductIds = angular.copy(blade.productIds);
+                items.query({ ids: blade.currentProductIds, respGroup: 'ItemInfo' }, function (data) {
+                    blade.$scope.gridApi.grid.options.data = data;
+                    blade.selectedListEntries = data.slice();
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: Do we need copying here?
Also, about angular.copy and .slice - probably it is better to use angular.copy in places we want to make copy (as it is deep copy). Slice creates shallow copy.

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                initialize();
+            };
+
+            $scope.cancel = function () {
+                initialize();
+            };
+
+            function initialize() {
+                blade.productIds = blade.promotion.productIds != undefined ? blade.promotion.productIds.slice() : [];
+                blade.currentProductIds = angular.copy(blade.productIds);
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: Do we really need 2 copies of items? Maybe we could have one copy as object for changes and use "blade.promotion.productIds" for comparison?

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+                    executeMethod: function () {
+                        var grid = blade.$scope.gridApi.grid;
+                        var selectedRows = grid.api.selection.getSelectedRows();
+                        for (var entry of selectedRows) {
+                            grid.options.data = _.filter(grid.options.data, function (item) {
+                                return item.id != entry.id;
+                            });
+                        }
+
+                        blade.selectedListEntries = grid.options.data.slice();
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: Do we need copy here?

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js (hunk 11 above)
Comment (blazemachine) [reply VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: This copy is needed for catalog items blade which could change it.

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js (hunk 13 above)
Comment (blazemachine) [reply VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: This copy is needed for catalog items blade which could change it.

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js [hunk 14]
```diff
@@ -61,13 +72,14 @@ angular.module('virtoCommerce.marketingModule')
...
-                    newBlade.error = undefined;
+                    catalogBlade.error = undefined;
                 }
             }
         };
 
-        bladeNavigationService.showBlade(newBlade, $scope.blade);
-    }
+        bladeNavigationService.showBlade(catalogBlade, $scope.blade);
+        return;
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js]: Do not need return here

**Comment by yecli** [comment]
@blazemachine Please attach demo gif with different cases of product selection (empty, 1 product with old productId, 1 product with productIds, >1 products selected).

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.tpl.html [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+                columnDefs: [
+                    { name: 'imgSrc', displayName: 'catalog.blades.catalog-items-select.labels.picture', cellTemplate: 'catalog-conditions-item-list-image-cell.tpl.html', enableColumnResizing: false, enableSorting: false, width: 60 },
+                    { name: 'name', displayName: 'catalog.blades.catalog-items-select.labels.name', enableSorting: false },
+                    { name: 'code', displayName: 'catalog.blades.catalog-items-select.labels.sku', width: 170, cellTemplate: 'catalog-conditions-item-list-SKU.cell.html', enableSorting: false }]})">
+                <div ui-grid="gridOptions" ui-grid-pinning ui-grid-auto-resize ui-grid-save-state ui-grid-selection ui-grid-resize-columns ui-grid-move-columns ui-grid-height></div>
+            </div>
+        </div>
+    </div>
+</div>
+<div class="blade-static __bottom">
```
Comment (tatarincev) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.tpl.html]: Need to use default Ok/Cancel widget instead https://github.com/VirtoCommerce/vc-module-store/blob/dev/VirtoCommerce.StoreModule.Web/Scripts/blades/taxProvider-detail.tpl.html#L1

File: VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+                    name: "platform.commands.add", icon: 'fa fa-plus',
+                    executeMethod: function () {
+                        openCatalogBlade();
+                    },
+                    canExecuteMethod: function () {
+                        return true;
+                    }
+                },
+                {
+                    name: "platform.commands.delete", icon: 'fa fa-trash-o',
```
Comment (tatarincev) [inline VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: See how to delete method implemented in marketing module https://github.com/VirtoCommerce/vc-module-marketing/blob/dev/VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-list.js#L44

Comment (yecli) [reply VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/catalog-entry-list.js]: Done

**Pull request merged**